### PR TITLE
GROOVY-7018 GroovyClassLoader addClasspath RegexPattern issue

### PR DESCRIPTION
**Pull request by Persi**

### CODE REVIEW

**Comment by PascalSchumacher** [comment]
Thanks for the pull request. :+1: 

Sadly the merge of this pull cause several test to fail:

**Comment by Persi** [comment]
Hello Pascal,

I've checked the failing testclass and it looks like this test is unable to run in Windows. In my Windows OS I get an error concerning the absolute path of the testfiles:

```
java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in opaque part at index 2: D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\target\resources\test
D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\src\test-resources\org\codehaus\groovy\ant\GroovycTest.xml:165: java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in opaque part at index 2: D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\target\resources\test
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:116)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.Target.execute(Target.java:435)
	at org.apache.tools.ant.Target.performTasks(Target.java:456)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1405)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1376)
	at org.codehaus.groovy.ant.GroovycTest.ensureExecutes(GroovycTest.java:234)
	at org.codehaus.groovy.ant.GroovycTest.ensureExecutesWithJavaHome(GroovycTest.java:240)
	at org.codehaus.groovy.ant.GroovycTest.testGroovycTest1_ForkGroovy_WithGroovyClasspath_WithJavaHome(GroovycTest.java:213)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at junit.framework.TestCase.runTest(TestCase.java:176)
	at junit.framework.TestCase.runBare(TestCase.java:141)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:86)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:49)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:64)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:106)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:360)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in opaque part at index 2: D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\target\resources\test
	at groovy.lang.GroovyClassLoader$6.run(GroovyClassLoader.java:940)
	at groovy.lang.GroovyClassLoader$6.run(GroovyClassLoader.java:906)
	at java.security.AccessController.doPrivileged(Native Method)
	at groovy.lang.GroovyClassLoader.addClasspath(GroovyClassLoader.java:906)
	at org.codehaus.groovy.ant.Groovyc.loadRegisteredScriptExtensions(Groovyc.java:1351)
	at org.codehaus.groovy.ant.Groovyc.execute(Groovyc.java:837)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:293)
	at sun.reflect.GeneratedMethodAccessor6.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	... 47 more
Caused by: java.net.URISyntaxException: Illegal character in opaque part at index 2: D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\target\resources\test
	at java.net.URI$Parser.fail(URI.java:2829)
	at java.net.URI$Parser.checkChars(URI.java:3002)
	at java.net.URI$Parser.parse(URI.java:3039)
	at java.net.URI.<init>(URI.java:595)
	at groovy.lang.GroovyClassLoader$6.run(GroovyClassLoader.java:922)
	... 57 more
```

 If I remove my changes in the exception handling of the GroovyClassLoader the test is OK ... but in fact occuring URISyntaxExceptions are just ignored. So I'm unsure if the test is the problem here or if my exception handling (which does not more then rethrowing occuring exceptions) is the problem. I would not like to remove the exception handling, because completely ignoring exceptions (especially in central classes like ClassLoaders!) is bad practice in my opinion.

The problem is that the exception above is not occuring in the testclass, it occurs in the `Project` class of the groovy-ant plugin.

So ... what shall we do? I could remove my exception rethrowing in the ClassLoader, which would not really solve the problem I think ... the test would be green but doesn't really test something on Windows machines. But the ClassLoader would not work correctly I believe.

Another way would be to just catch and ignore occuring exceptions in the ClassLoader, but log them into STD_OUT. So a user would have the chance to find false pathes in his classpath, otherwise he got no chance to find out, why a defined path is ignored by Groovy.

**Comment by Persi** [comment]
Hello Pascal,

The occured testerrors just looks like missing files in the test environment. But I'll try to reproduce them in a Linux VM soon.

In the moment I can't reproduce the occured exceptions on my Windows platform because of a likely bug in ant on Windows machines which causes a URISyntaxException in the GroovyClassLoader:

```
java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in opaque part at index 2: D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\target\resources\test
D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\src\test-resources\org\codehaus\groovy\ant\GroovycTest.xml:165: java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in opaque part at index 2: D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\target\resources\test
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:116)
	at org.apache.tools.ant.Task.perform(Task.java:348)
	at org.apache.tools.ant.Target.execute(Target.java:435)
	at org.apache.tools.ant.Target.performTasks(Target.java:456)
	at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1405)
	at org.apache.tools.ant.Project.executeTarget(Project.java:1376)
	at org.codehaus.groovy.ant.GroovycTest.ensureExecutes(GroovycTest.java:234)
	at org.codehaus.groovy.ant.GroovycTest.ensureExecutesWithJavaHome(GroovycTest.java:240)
	at org.codehaus.groovy.ant.GroovycTest.testGroovycTest1_ForkGroovy_WithGroovyClasspath_WithJavaHome(GroovycTest.java:213)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at junit.framework.TestCase.runTest(TestCase.java:176)
	at junit.framework.TestCase.runBare(TestCase.java:141)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:86)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:49)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:64)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:106)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:360)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in opaque part at index 2: D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\target\resources\test
	at groovy.lang.GroovyClassLoader$6.run(GroovyClassLoader.java:940)
	at groovy.lang.GroovyClassLoader$6.run(GroovyClassLoader.java:906)
	at java.security.AccessController.doPrivileged(Native Method)
	at groovy.lang.GroovyClassLoader.addClasspath(GroovyClassLoader.java:906)
	at org.codehaus.groovy.ant.Groovyc.loadRegisteredScriptExtensions(Groovyc.java:1351)
	at org.codehaus.groovy.ant.Groovyc.execute(Groovyc.java:837)
	at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:293)
	at sun.reflect.GeneratedMethodAccessor6.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
	... 47 more
Caused by: java.net.URISyntaxException: Illegal character in opaque part at index 2: D:\gitworkspace\incubator-groovy\subprojects\groovy-ant\target\resources\test
	at java.net.URI$Parser.fail(URI.java:2829)
	at java.net.URI$Parser.checkChars(URI.java:3002)
	at java.net.URI$Parser.parse(URI.java:3039)
	at java.net.URI.<init>(URI.java:595)
	at groovy.lang.GroovyClassLoader$6.run(GroovyClassLoader.java:922)
	... 57 more
```

**Comment by Persi** [comment]
Hello Pascal,

I've added a fix for the failing testclass.

**Comment by Persi** [comment]
As I see now a Gradle task failed because of a false Classpathconfiguration. Taking a look at it ...

**Comment by Persi** [comment]
Now Teamcity looks good :)

**Comment by PascalSchumacher** [comment]
Nice :)

Thanks for all your effort. :+1: 

It looks like the changes to the test are not necessary. I was able to run the test locally on windows without the changes and a pull request without the test changes runs on the buildserver: http://ci.groovy-lang.org/viewLog.html?buildId=24483&tab=buildResultsDiv&buildTypeId=Groovy_Jdk7Build

It would be nice if you could revert the changes to the tests and squash all commits.

I will then try to motivate a member of the groovy team more knowledgeable about this code to do a review. Hopefully we can merge this pull request soon.

Thanks,
Pascal

**Comment by Persi** [comment]
Hello Pascal,

after fighting some time with github (I'm more or less new to it ...) I squashed my commits to one.

Greetings,
Marcus

**Comment by PascalSchumacher** [comment]
Hi Marcus,

great :).

I used to struggle a lot with git/github (still do sometimes), it's a 
lot to learn.

I just sent this mail to the developer mailing list:

-------- Weitergeleitete Nachricht --------
Betreff: 	GROOVY-7018 GroovyClassLoader addClasspath RegexPattern issue 
/ Review of pull request 43
Datum: 	Tue, 30 Jun 2015 22:44:53 +0200
Von: 	Pascal Schumacher
Antwort an: 	dev@groovy.incubator.apache.org
An: 	dev@groovy.incubator.apache.org



Hi everybody,

multiple people reported to be affected by this regression
(https://issues.apache.org/jira/browse/GROOVY-7018) and Jochen scheduled
it to be fixed in multiple 2.3.X versions, but we never got around to
actually fix it.

Now we got a pull request
(https://github.com/apache/incubator-groovy/pull/43) which passes all
tests on the ci server
(http://ci.groovy-lang.org/viewLog.html?buildId=24487&tab=buildResultsDiv&buildTypeId=Groovy_Jdk7Build)
and looks o.k. to me.

It would be very nice, if anybody more knowledgeable in this area could
do review.

Thanks and kind regards,
Pascal

File: src/main/groovy/ui/GroovyMain.java [hunk 1]
```diff
@@ -391,17 +391,16 @@ public String getText(String uriOrFilename) throws IOException {
...
+            // search for the file and if it exists don't try to use URIs ...
+            File scriptFile = huntForTheScriptFile(script);
+            if (!scriptFile.exists() && uriPattern.matcher(script).matches()) {
                 return new GroovyCodeSource(new URI(script));
-            } else {
-                return new GroovyCodeSource(huntForTheScriptFile(script));
             }
-        } else {
-            return new GroovyCodeSource(script, "script_from_command_line", GroovyShell.DEFAULT_CODE_BASE);
+            return new GroovyCodeSource(huntForTheScriptFile(script));
```
Comment (paulk-asert) [inline src/main/groovy/ui/GroovyMain.java]: can't you just use "scriptFile" instead of the second call to "huntForTheScriptFile"?

**Comment by blackdrag** [comment]
I actually wonder about if we try to avoid using the pattern by building on top of the exception mechanism in GroovyClassLoader, then why don't we do the same logic in GroovyMain?

File: src/main/groovy/ui/GroovyMain.java (hunk 1 above)
Comment (Persi) [inline src/main/groovy/ui/GroovyMain.java]: Hello Paul,

indeed I could reduce the calls to "huntForTheScriptFile" to one.

I'll commit it.

**Comment by Persi** [comment]
Hello blackdrag,

that's what i thought about as well.

In fact one reason for me is, that in GroovyMain the given "script" String has to be an existing file if it is a scriptfile. In the ClassLoader it is possible to write pathes in the classpath which doesn't actually exist, but could exist later. That's why we can use file.exists() in GroovyMain, while this check isn't working for any cases in the ClassLoader.

Another reason is, that the defined URI pattern is completely RFC conform and it just fails in Cygwin environments if the given script is created with a false cygpath call. This problem is catched by using the file instead of the URI. But if we got a correct URI in GroovyMain (the only way to determine if it's correct, if we ignore this exception driven development I had to implement in the ClassLoader, is to use the pattern) we should use this URI as is.

I'm not sure which way is the better one. My solution in the GroovyClassloader should also work in GroovyMain, but if we would use it, we couldn't remove the URI pattern because it is already used in `public String getText(String uriOrFilename)` as well.

Greets,
Marcus

**Comment by PascalSchumacher** [comment]
@blackdrag @paulk-asert Is this good to merge?

**Comment by blackdrag** [comment]
+1

**Comment by PascalSchumacher** [comment]
@Persi  Thanks for the pull request. :+1: 

@blackdrag @paulk-asert Thanks for the reviews. :)

**Comment by Persi** [comment]
Thanks for bringing my PR into production that fast! :)

**Pull request closed without merging**